### PR TITLE
[WIP] Add unit test coverage to existing behavior

### PR DESCRIPTION
**Pull request by jeremiak**
**This is a work in progress and should NOT be merged**

Working on pinning existing behavior with some JS unit tests with @toolness

Some open questions I have:
- Are we actually testing the script behavior in a good way? I mean right now we're just making sure the commands get executed appropriately, but perhaps we should also be testing that the end result is achieved?

### CODE REVIEW

**Comment by toolness** [comment]
Wooot!! This is looking good.

Another option to the `console.log`'ing of the "stub" scripts may be for them to append to some kind of JSON log file, but that's probably only needed if you actually run into problems w/ doing things the b64 way.

Regarding testing that the end result is achieved--that's a good idea... I wonder if, for `clone.sh`, it makes sense to have a different `git` stub that *mostly* passes its arguments off to the *real* `git` command, but modifies any HTTPS URLs to `github.com` and instead makes them `file:` URLs to the local filesystem?  (If only I knew as much about bash as I do about other languages, I'd say that perhaps you could even refactor the clone script to have a `generateGithubURL()` function that gets monkey-patched by your test suite, but I don't know how to do that, or if it's even possible.)

Anyways, that way you could test the code out on a "real" git repository, and even have different git repositories that use different static site generators and such... Could potentially be useful for lots of different kinds of integration testing.  (Or, alternatively, you could even set up a local git server that is communicated with over HTTP, but I don't know much about how to do that.)

Similarly for uploading to S3, I wonder if you could use something like [s3rver](https://www.npmjs.com/package/s3rver) to provide a fake s3 environment that you could inspect before and after running the scripts?

File: test/clone.js [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+    '--single-branch',
+    `https://@github.com/${OWNER}/${REPOSITORY}.git`,
+    '.'
+  ];
+
+  t.plan(1);
+  t.deepEqual(command, expectedCloneArguments);
+});
+
+test('executes clone command with SOURCE_OWNER or SOURCE_REPO', (t) => {
```
Comment (brandocalrissian) [inline test/clone.js]: Does it make sense to handle all of the cases (has SOURCE_OWNER but not SOURCE_REPO, vice versa and both)?

**Comment by brandocalrissian** [comment]
This is really intense. It's almost "golden file" testing as we're basically looking to match the command line parameters in our test. However, the test looks pretty resilient to catching changes to clone.sh. How could we make it less brittle? The best I can come up with is to add a layer of indirection:

 * make the script execute your application code and test that it's called correctly (the right environment variables get wired to the call)
 * encapsulate the particular shelling out commands to something tied directly to the particular command (e.g. git or rm) and test these things create the right parameters to the command line tool
 * verify the application code takes its properties and orchestrates the correct calls to the command line encapsulation jobbies built previously

You could test-drive this development and potentially do golden-file integration tests to verify the functionality doesn't change from what you have today (and then drop the brittle golden-file tests when they stop helping).

A reasonable person could say this would be extreme overkill.

**Comment by brandocalrissian** [comment]
The fact this really tests what's going on is fantastic!

**Comment by wslack** [comment]
@jeremiak and @toolness what happened with this?

**Comment by toolness** [comment]
Hi!  I'm happy to continue helping out when I can.  I wasn't sure what capacity @jeremiak has on it since I thought he's on the FBI project now?

**Comment by wslack** [comment]
Yeah, IIRC if was part of the mentorship work, not directly part of Federalist. Just don't want to lose coordination.

**Comment by wslack** [comment]
@toolness @jeremiak I am moving this back to backlog unless y'all let me know otherwise; thanks!

**Comment by wslack** [comment]
@jeremiak should we just close this?

**Comment by jeremiak** [comment]
Fine by me!

**Pull request closed without merging**